python/ray/util/helpers.py

```diff
@@ -0,0 +1,249 @@
+
+    # num_returns must be <= len(refs)
+    ready, refs = ray.wait(
+        refs,
+        num_returns=min(chunk_size, len(refs)),
+        **kwargs,
+    )
+
+    if not return_objrefs:
+        return ray.get(ready), refs
+
+    return ready, refs
+
+
+@PublicAPI(stability="alpha")
+def as_completed(
+    refs: "Sequence[ObjectRef]",
+    *,
+    chunk_size: int = DEFAULT_CHUNK_SIZE,
+    return_objrefs: bool = False,
+    **kwargs,
+) -> Iterator[Union[Any, "ObjectRef"]]:
+    """Given a list of Ray task references, yield results as they become available.
+
+    Unlike calling :meth:`~ray.get` on a list of references (i.e., `ray.get(refs)`) which
+    waits for all results to be ready, this function begins to yield result as soon as
+    a batch of `chunk_size` results are ready.
+
+    .. note::
+        Generally there is no guarantee on the order of results. For example, the first result
+        is not necessarily the first one completed, but rather the first one submitted in the
+        first available batch (See :meth:`~ray.wait` for more details about
+        preservation of submission order).
+
+    .. note::
+        Use this function instead of calling :meth:`~ray.get` inside a for loop. See
+        https://docs.ray.io/en/latest/ray-core/patterns/ray-get-loop.html for more details.
+
+    Example:
+        Suppose we have a function that sleeps for x seconds depending on the input.
+        We expect to obtain a partially sorted list of results.
+
+        .. testcode:: python
+            import ray
+            import time
+
+            @ray.remote
+            def f(x):
+                time.sleep(x)
+                return x
+
+            refs = [f.remote(i) for i in [10, 4, 6, 8, 2]]
+            for x in ray.util.as_completed(refs, chunk_size=2):
+                print(x)
+
+        .. testoutput::
+            :options: +MOCK
+
+            # Output:
+            4
+            2
+            6
+            8
+            10
+
+    Args:
+        refs: A list of Ray object refs.
+        chunk_size: The number of tasks to wait for in each iteration (default 10).
+            The parameter is passed as `num_returns` to :meth:`~ray.wait` internally.
+        return_objrefs: If True, return Ray remote refs instead of results (by calling :meth:`~ray.get`).
+        **kwargs: Additional keyword arguments to pass to :meth:`~ray.wait`, e.g.,
+            `timeout` and `fetch_local`.
+
+    Yields:
+        The results (or optionally their Ray references) of the Ray tasks as they complete.
+    """
+    if chunk_size < 1:
+        raise ValueError("`chunk_size` must be >= 1")
+
+    if "num_returns" in kwargs:
+        raise ValueError("Use the `chunksize` argument instead of `num_returns`.")
+
+    while refs:
+        results, refs = _wait_and_get_single_batch(
+            refs,
+            chunk_size=chunk_size,
+            return_objrefs=return_objrefs,
+            **kwargs,
+        )
+        yield from results
```

Comment:
wow, after years of wrestling with python generators, you've taught me something new :)